File: data_structures_and_algorithms/algorithms/sort/Selection.java
package data_structures_and_algorithms.algorithms.sort;

import edu.princeton.cs.algs4.StdIn;
import edu.princeton.cs.algs4.StdOut;

public class Selection {
    private Selection() {

    }

    public static void sort(Comparable[] a) {
        int n = a.length;
        for (int i = 0; i < n; i++) {
            int min = i;
            for (int j = i + 1; j < n; j++) {
                if (less(a[j], a[min])) {
                    min = j;
                }
            }
            exch(a, i, min);
            assert isSorted(a, 0, i);

        }
        assert isSorted(a);
    }

    private static boolean less(Comparable v, Comparable w) {
        // 小于就返回负数，所以小于返回 true，也就是 < 0
        return v.compareTo(w) < 0;
    }

    // exchange a[i] and a[j]
    private static void exch(Object[] a, int i, int j) {
        Object swap = a[i];
        a[i] = a[j];
        a[j] = swap;
    }

    private static boolean isSorted(Comparable[] a, int lo, int hi) {
        for (int i = lo + 1; i <= hi; i++)
            if (less(a[i], a[i - 1]))
                return false;
        return true;
    }

    // is the array a[] sorted?
    private static boolean isSorted(Comparable[] a) {
        return isSorted(a, 0, a.length - 1);
    }

    private static void show(Comparable[] a) {
        for (int i = 0; i < a.length; i++) {
            StdOut.println(a[i]);
        }
    }

    public static void main(String[] args) {
        String[] a = StdIn.readAllStrings();
        Selection.sort(a);
        show(a);
    }

}

File: data_structures_and_algorithms/algorithms/sort/selection.py
class Selection:

    def sort(self, list: list):
        length: int = len(list)
        for i in range(length):
            min: int = i
            for j in range(i, length):
                if self.less(list, j, min):
                    min = j
            self.exch(list, i, min)

    def less(self, list: list, i: int, j: int) -> bool:
        return list[i] < list[j]

    def exch(self, list: list, i: int, j: int) -> None:
        list[i], list[j] = list[j], list[i]

    def show(self, list: list) -> None:
        for i in list:
            print(i)

    def _main(self):
        test_list = [1, 3, 2, 10, 8, 7, 20, 1]
        self.sort(test_list)
        self.show(test_list)


if __name__ == "__main__":
    Selection()._main()
